README


This archive contains the distribution Net-Async-PostgreSQL,
version 0.006:

  PostgreSQL database support for IO::Async

This software is copyright (c) 2011 by Tom Molesworth.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
0.006     2011-10-28 04:19:11 Europe/London
	Set up local PostgreSQL instance via Test::postgresql so we can run proper tests against
	a server rather than just checking method handling basics. Big thanks to Kazuho Oku for
	providing the Test::postgresql module, and Matt Trout for directing me to it.

	No functional changes here.

0.005     2011-07-16 22:43:10 Europe/London
	Support new event handling brought in by Protocol::PostgreSQL 0.008
	Moved Protocol::PostgreSQL::Client into an attribute rather than an ancestor
	of our class, so we now proxy most of the methods through including
	->add_handler_for_event

0.004     2011-06-11 17:44:06 Europe/London
	Pass through service parameter so that we can connect to ports other than the default 5432.
	Provide a ->terminate method for clean shutdown.

0.003     2011-02-26 19:01:57 Europe/London
	Update synopsis to use ->configure and increase minimum version requirement.

0.002     2011-02-22 20:49:19 Europe/London
	Fix a nasty packet-boundary bug and chuck in a synopsis.

0.001     2011-02-12 12:44:52 Europe/London
	Preliminary release to CPAN. Only supports the basic query interface and no type handling
	yet, but allows basic queries and connection via plaintext or MD5.



examples/client.pl
#!/usr/bin/perl
use strict;
use warnings;

use IO::Async::Loop;
use Net::Async::PostgreSQL::Client;

my $loop = IO::Async::Loop->new;
my $client = Net::Async::PostgreSQL::Client->new(
	debug			=> 0,
	host			=> $ENV{NET_ASYNC_POSTGRESQL_SERVER} || 'localhost',
	service			=> $ENV{NET_ASYNC_POSTGRESQL_PORT} || 5432,
	database		=> $ENV{NET_ASYNC_POSTGRESQL_DATABASE},
	user			=> $ENV{NET_ASYNC_POSTGRESQL_USER},
	pass			=> $ENV{NET_ASYNC_POSTGRESQL_PASS},
);
#$client->init;

my @query_list = (
	q{begin work},
	q{create schema nap_test},
	q{create table nap_test.nap_1 (id serial primary key, name varchar, creation timestamp)},
	q{insert into nap_test.nap_1 (name, creation) values ('test', 'now')},
	q{insert into nap_test.nap_1 (name, creation) values ('test2', 'now')},
	q{select * from nap_test.nap_1},
);
my $init = 0;
my $finished = 0;
my %status;
$client->attach_event(
	error	=> sub {
		my ($self, %args) = @_;
		print "Received error\n";
		my $err = $args{error};
		warn "$_ => " . $err->{$_} . "\n" for sort keys %$err;
	},
	command_complete => sub {
		my $self = shift;
		print "Command complete\n";
		warn $finished;
		$loop->loop_stop if $finished == 2;
	},
	copy_in_response => sub {
		my ($self, %args) = @_;
		print "Copy in response\n";
		$self->copy_data("some name\t2010-01-01 00:00:00");
		++$finished;
		$self->copy_done;
	},
	ready_for_query => sub {
		my $self = shift;
		print "Ready for query\n";
		unless($init) {
			print "Server version " . $status{server_version} . "\n";
			++$init;
		}
		my $q = shift(@query_list);
		if($finished == 1) {
			print "run query\n";
			$self->simple_query(q{select * from nap_test.nap_1});
			++$finished;
			return;
		} elsif($finished == 2) {
			$loop->loop_stop;
		}

		if($q) {
			$self->simple_query($q);
		} else {
			$self->simple_query(q{copy nap_test.nap_1 (name,creation) from stdin});
		}
	},
	parameter_status => sub {
		my $self = shift;
		print "Parameter status\n";
		my %args = @_;
		$status{$_} = $args{status}->{$_} for sort keys %{$args{status}};
	},
	row_description => sub {
		my $self = shift;
		print "Row description\n";
		my %args = @_;
		print '[' . join(' ', map { $_->{name} } @{$args{description}{field}}) . "]\n";
	},
	data_row => sub {
		my $self = shift;
		print "Data row\n";
		my %args = @_;
		print '[' . join(',', map { $_->{data} } @{$args{row}}) . "]\n";
	}
);
$loop->add($client);
$client->connect;
$loop->loop_forever;
exit 0;


examples/benchmark.pl
#!/usr/bin/perl
use strict;
use warnings;
use Benchmark qw(:hireswallclock);

use IO::Async::Loop;
use Net::Async::PostgreSQL::Client;

my $code = sub {
	my $port = shift;
my $loop = IO::Async::Loop->new;
my $client = Net::Async::PostgreSQL::Client->new(
	debug			=> 0 && sub { warn "@_" },
	host			=> $ENV{NET_ASYNC_POSTGRESQL_SERVER} || 'localhost',
	service			=> $port,
	ssl			=> 1,
	database		=> $ENV{NET_ASYNC_POSTGRESQL_DATABASE},
	user			=> $ENV{NET_ASYNC_POSTGRESQL_USER},
	pass			=> $ENV{NET_ASYNC_POSTGRESQL_PASS},
);
#$client->init;

my @query_list = (
	q{begin work},
	q{create schema nap_test},
	q{create table nap_test.nap_1 (id serial primary key, name varchar, creation timestamp, id1 bigint, id2 bigint, k1 text, k2 text)},
);

my $init = 0;
my $finished = 0;
my %status;
my $rfq = sub {
	my $self = shift;
	unless($init) {
#		print "Server version " . $status{server_version} . "\n";
		++$init;
	}
	my $q = shift(@query_list);
	return $self->simple_query($q) if $q;

	if($finished == 1) {
		$self->simple_query(q{select * from nap_test.nap_1});
		++$finished;
		return;
	} elsif($finished == 2) {
		$self->simple_query(q{rollback});
		++$finished;
		return;
	} elsif($finished == 3) {
		$loop->later(sub {
			$self->transport->configure(on_outgoing_empty => sub {
				$self->close;
				$loop->later(sub { $loop->loop_stop; });
			});
			$self->send_message('Terminate');
		});
		++$finished;
		return;
	}

	$self->simple_query(q{copy nap_test.nap_1 (name,creation, id1, id2, k1, k2) from stdin});
	return;
};
$client->attach_event(
	error	=> sub {
		my ($self, %args) = @_;
		print "Received error\n";
		my $err = $args{error};
		warn "$_ => " . $err->{$_} . "\n" for sort keys %$err;
	},
	command_complete => sub {
		my $self = shift;
	},
	copy_in_response => sub {
		my ($self, %args) = @_;
		my $k1 = 'aaaaaaaaaaa';
		$self->send_copy_data(['some name', '2010-01-01 00:00:00', int(rand() * 10_000_000), $$ ^ time, ++$k1, localtime() . {} ]) for 0..100_000;
		++$finished;
		$self->copy_done;
#		$self->attach_event(ready_for_query => $rfq);
	},
	ready_for_query => $rfq,
	parameter_status => sub {
		my $self = shift;
		my %args = @_;
		$status{$_} = $args{status}->{$_} for sort keys %{$args{status}};
	},
	row_description => sub {
		my $self = shift;
		my %args = @_;
	},
	data_row => sub {
		my $self = shift;
		my %args = @_;
	},
);
$loop->add($client);
$client->connect;
$loop->loop_forever;
};
timethese(1, {
	5432	=> sub { $code->(5432, @_) },
	6432	=> sub { $code->(6432, @_) }
});

exit 0;


lib/Net/Async/PostgreSQL.pm
package Net::Async::PostgreSQL;
# ABSTRACT: PostgreSQL database support for IO::Async
use strict;
use warnings;
our $VERSION = '0.006';

=head1 NAME

Net::Async::PostgreSQL - (preliminary) asynchronous PostgreSQL support for L<IO::Async>

=head1 VERSION

version 0.006

=head1 SYNOPSIS

 use strict; use warnings;
 use IO::Async::Loop;
 use Net::Async::PostgreSQL::Client;

 my $loop = IO::Async::Loop->new;
 my $client = Net::Async::PostgreSQL::Client->new(
	host			=> $ENV{NET_ASYNC_POSTGRESQL_SERVER} || 'localhost',
	service			=> $ENV{NET_ASYNC_POSTGRESQL_PORT} || 5432,
	database		=> $ENV{NET_ASYNC_POSTGRESQL_DATABASE},
	user			=> $ENV{NET_ASYNC_POSTGRESQL_USER},
	pass			=> $ENV{NET_ASYNC_POSTGRESQL_PASS},
 );
 $client->init;

 $client->configure(
 	on_error	=> sub {
 		my ($self, %args) = @_;
 		my $err = $args{error};
 		warn "$_ => " . $err->{$_} . "\n" for sort keys %$err;
 	},
 	on_ready_for_query => sub {
 		my $self = shift;
 		unless($init) {
 			print "Server version " . $status{server_version} . "\n";
 			++$init;
 		}
                 unless(keys %sth) {
                         $self->simple_query(q{begin work});
                         my $sth = $self->prepare_async(
                                 sql => q{insert into sometable(name) values ($1) returning idsometable, name},
 				statement => 'new_value',
 			);
                 }
 		$sth->bind("some more data");
 	},
 	on_parameter_status => sub {
 		my $self = shift;
 		my %args = @_;
 		$status{$_} = $args{status}->{$_} for sort keys %{$args{status}};
 	},
 	on_row_description => sub {
 		my $self = shift;
 		my %args = @_;
 		print '[' . join(' ', map { $_->{name} } @{$args{description}{field}}) . "]\n";
 	},
 	on_data_row => sub {
 		my $self = shift;
 		my %args = @_;
 		print '[' . join(',', map { $_->{data} } @{$args{row}}) . "]\n";
 		$loop->loop_stop;
 	}
 );
 $loop->add($client);
 $client->connect;
 $loop->loop_forever;

=head1 DESCRIPTION

See L<Net::Async::PostgreSQL::Client> or L<Protocol::PostgreSQL>.

=cut

1;

__END__

=head1 AUTHOR

Tom Molesworth <cpan@entitymodel.com>

=head1 LICENSE

Copyright Tom Molesworth 2011. Licensed under the same terms as Perl itself.


lib/Net/Async/PostgreSQL/Client.pm
package Net::Async::PostgreSQL::Client;
{
  $Net::Async::PostgreSQL::Client::VERSION = '0.006';
}
use strict;
use warnings;
use Protocol::PostgreSQL::Client '0.008';
use parent qw{IO::Async::Protocol::Stream};
use Scalar::Util ();

=head1 NAME

Net::Async::PostgreSQL - support for the PostgreSQL wire protocol

=head1 VERSION

version 0.006

=head1 SYNOPSIS

 # Simple queries are performed similar to DBI:
 $dbh->do(q{insert into something (x,y,z) values (1,2,3)});

 # These can also use bind variables:
 $dbh->do(q{insert into something (x,y,z) values (?,?,?)}, undef, 1,2,3);

 # Prepared statements work the same as DBI by default
 my $sth = $dbh->prepare(q{select * from table where name = ?});
 $sth->bind_param(1, 'test');
 $sth->execute;

 # ... but have async_ versions for passing handlers:
 my $sth = $dbh->async_prepare(
 	sql => q{select * from table where name = ?},
	on_error => sub { warn "failed" }
 );
 $sth->async_execute(
 	on_bind_request => sub {
		return @param;
	},
	on_header	=> sub { ... },
	on_row		=> sub { ... },
	on_error	=> sub { ... },
	on_complete	=> sub { ... },
 );

 # And there's a helper method for doing regular queries:
 $dbh->run_query(
 	sql		=> q{select * from something where id = ?},
	parameters	=> [1],
	on_row		=> sub { warn "Had " . $_[1]->{} },
	on_error	=> sub { warn "Error encountered" },
	on_complete	=> sub { warn "all done" }
 );

=head1 DESCRIPTION

The interface is provided by L<Net::Async::DBI>, which attempts to offer something close to
L<DBI> but with support for event-based request handling.

See L<Protocol::PostgreSQL> for more details.

=cut

use Socket qw(SOCK_STREAM);

BEGIN {
	foreach my $k (qw(
		send_message
		has_queued
		is_authenticated
		is_first_message
		initial_request
		queue
		send_next_in_queue
		message
		debug
		handle_message
		message_length
		simple_query
		copy_data
		copy_done
		backend_state
		active_statement
		prepare
		prepare_async
		row_description
		is_ready
		send_copy_data
		add_handler_for_event
	)) {
		no strict 'refs';
		*{join '::', __PACKAGE__, $k} = sub { shift->pg->$k(@_) };
	}
}

=head1 METHODS

=cut

sub new {
	my $class = shift;
	my %args = @_;

# Clear any options that will cause the parent class to complain
	my $loop = delete $args{loop};

# Want the IO::Async::Protocol constructor, so SUPER is good enough for us here
	my $self = $class->SUPER::new;
	$self->pg->add_handler_for_event(send_request => $self->_capture_weakself(sub {
		my ($self) = splice @_, 0, 2; # ignore pg object
		$self->write(@_);
		return 1;
	}));
	$self->configure(%args);

# Automatically add to the event loop if we were passed one
	$loop->add($self) if $loop;
	return $self;
}

sub sap {
	my ($self, $sub) = @_;
	Scalar::Util::weaken $self;
	return sub { $self->$sub(@_); };
}

=head2 configure

Apply callbacks and other parameters, preparing state for event loop start.

=cut

sub configure {
	my $self = shift;
	my %args = @_;

# Debug flag is used to control the copious amounts of data that we dump out when tracing
	$self->{debug} = $args{debug} if exists $args{debug};

#	%args = $self->pg->configure(%args);
	foreach (qw{host service}) {
		$self->{$_} = delete $args{$_} if exists $args{$_};
	}
	%args = $self->pg->configure(%args);
	$self->SUPER::configure(%args);
	return $self;
}

sub pg {
	my $self = shift;
	if(@_) {
		$self->{pg} = shift;
		return $self;
	}
	$self->{pg} ||= Protocol::PostgreSQL::Client->new;
	return $self->{pg};
}

=head2 on_connection_established

Prepare and activate a new transport.

=cut

sub on_connection_established {
	my $self = shift;
	my $sock = shift;
	my $transport = IO::Async::Stream->new(handle => $sock)
		or die "No transport?";
	$self->configure(transport => $transport);
	$self->debug("Have transport " . $self->transport);
}

=head2 on_starttls

Upgrade the underlying stream to use TLS.

=cut

sub on_starttls {
	my $self = shift;
	$self->debug("Upgrading to TLS");

	require IO::Async::SSLStream;

	$self->SSL_upgrade(
		on_upgraded => $self->_capture_weakself(sub {
			my ($self) = @_;
			$self->debug("TLS upgrade complete");

			$self->{tls_enabled} = 1;
			$self->initial_request;
		}),
		on_error => sub { die "error @_"; }
	);
}

=head2 connect

=cut

sub connect {
	my $self = shift;
	my %args = @_;

	my $on_connected = delete $args{on_connected};
	my $host = exists $args{host} ? delete $args{host} : $self->{host};
	$self->pg->add_handler_for_event(password => sub {
		my $self = shift;
		$self->send_message('PasswordMessage', password => $self->{pass});
		return 0; # single-shot event
	});
	$self->SUPER::connect(
		%args,
		service		=> $args{service} || $self->{service} || 5432,
		host		=> $host,
		socktype	=> SOCK_STREAM,
		on_resolve_error => sub {
			die "Resolution failed for $host";
		},
		on_connect_error => sub {
			die "Could not connect to $host";
		},
		on_connected => $self->sap(sub {
			my $self = shift;
			my ($pg, $sock) = @_;
			$self->pg->initial_request;
			$on_connected->($self) if $on_connected;
		})
	);
}

=head2 on_read

Handle read requests by passing full packets back to the protocol handler.

=cut

sub on_read {
	my $self = shift;
	my ($buffref, $eof) = @_;
	return 0 unless length($$buffref) >= 5;
	my ($code, $size) = unpack('C1N1', $$buffref);
	if(length($$buffref) >= $size+1) {
		$self->pg->handle_message(substr $$buffref, 0, $size+1, '');
		return 1;
	}
	return 0;
}

=head2 do


=cut

sub do {
	my $self = shift;
	my ($sql, $attrib, @param) = @_;
	$self->simple_query($sql);
	return $self;
}

sub on_password {
	my $self = shift;
}

=head2 terminate

Sends the Terminate message to the database server and closes the connection for a clean
shutdown.

=cut

sub terminate {
	my $self = shift;
	return unless $self->transport;

	my $transport = $self->transport;
	Scalar::Util::weaken(my $loop = $transport->get_loop);
	# TODO could just ->close_when_empty?
	$transport->configure(on_outgoing_empty => $self->_capture_weakself(sub {
		my $self = shift;
		$self->close;
		$loop->later($self->_capture_weakself(sub { $self->pg->invoke_event('closed'); }));
	}));
	$self->send_message('Terminate');
	return $self;
}

1;

__END__

=head1 SEE ALSO

=over 4

=item * L<DBI> - the real database interface

=item * L<DBD::Gofer> - proxy request support for DBI

=back

=head1 AUTHOR

Tom Molesworth <cpan@entitymodel.com>

=head1 LICENSE

Copyright Tom Molesworth 2011. Licensed under the same terms as Perl itself.


Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "PostgreSQL database support for IO::Async",
  "AUTHOR" => "Tom Molesworth <cpan\@entitymodel.com>",
  "BUILD_REQUIRES" => {
    "Test::postgresql" => "0.09"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Net-Async-PostgreSQL",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Net::Async::PostgreSQL",
  "PREREQ_PM" => {
    "IO::Async" => "0.36",
    "Protocol::PostgreSQL" => "0.008",
    "parent" => 0
  },
  "VERSION" => "0.006",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



